src/format/Format.cc

```diff
@@ -472,15 +472,9 @@ Format::Format::assemble(MemBuf &mb, const AccessLogEntry::Pointer &al, int logS
 
         case LFT_LOCAL_LISTENING_IP: {
             // avoid logging a dash if we have reliable info
-            const bool interceptedAtKnownPort = al->request ?
-                                                (al->request->flags.interceptTproxy ||
-                                                 al->request->flags.intercepted) && al->cache.port :
-                                                false;
-            if (interceptedAtKnownPort) {
-                const bool portAddressConfigured = !al->cache.port->s.isAnyAddr();
-                if (portAddressConfigured)
```

Comment:
The removal of this cache.port use as a backup data source also removes existing ability to log and use details of the *_port config settings when there is no HTTP request on a connection. Specifically for SSL-Bump errors and TLS/SSL helper key_extras format codes, ERR_INVALID_REQ, ERR_UNSUP_HTTPVERSION, ERR_UNSUP_REQ, ERR_PROTOCOL_UNKNOWN, and ERR_LIFETIME_EXP.

Reply:
Added ability to pass al->cache.port->s and al->tcpClient->local as parameters. They will take precedence over analogues IPs obtainable via request->masterXaction. This should resolve above issues. Please check.

Reply:
I see two mutually exclusive paths forward:

1. Prove that `masterXaction->squidPort` is equivalent to `al->cache.port`. Remove the recently added (and hard to compute correctly) FindListeningPortAddress() parameters. Restore the previously removed `al->tcpClient` check here.
2. Remove the recently added (and hard to compute correctly) FindListeningPortAddress() parameters. Add a `const AccessLogEntryPointer &ale` parameter. Reproduce the original LFT_LOCAL_LISTENING_IP code inside FindListeningPortAddress() to restore the original functionality "as is" (when a non-nil `ale` is passed) and then add your code that uses masterXaction->squidPort for cases when `ale` is nil.

If `masterXaction->squidPort` is equivalent to `al->cache.port` (when both are available), then the first option would result in overall better code. If the two may differ in some cases, then the second option allows you to avoid fixing that bug while still gaining the functionality you want.

Reply:
In my opinion my new code FindListeningPortAddress() is much better, cleaner and re-useable, although it looks complex due to a complete rewrite of original logic. (I will try to explain it below)

First of all personally its impossible for me to go through whole code and prove that masterXaction->squidPort is equivalent to al->cache.port.

I have picked up hints from Amos here:
http://lists.squid-cache.org/pipermail/squid-users/2018-May/018190.html

> The Convert() equivalent of "al->request" is just "request".
> The Convert() equivalent of "al->tcpClient" is "request->masterXaction->tcpClient".
> The Convert() equivalent of "al->cache.port" is "request->masterXaction->squidPort".

# Explanation:

There are two cases where FindListeningPortAddress() would be used. One when dealing with ALE (File: Format.cc) and other is in case of deny_info (or error pages) (File: errorpage.cc)

## **Case 1:**
If using ALE (al->) then code uses al->request to first detect if its intercepted.

If intercepted then it returns al->cache.port->s (which is passed as cacheip).(This satisfies Amos's review / issue raised by him)

If not intercepted then it returns al->tcpClient->local (which is passed as localip). This eliminates `else if (al->tcpClient)` part of Format.cc

I am assuming that if information is NOT available to al-> then it wont be available to request->masterXaction too. (because whole discussion above we are talking about something available via al->X but al->request may still be NULL)

So: (I expect that)
if al->cache.port is NULL then request->masterXaction->squidPort would be NULL too.
if al->tcpClient is NULL then request->masterXaction->tcpClient would be NULL too.

## **Case 2:**
If we are not dealing with ALE but request pointer itself then cacheip and localip are passed as nullptr. Code then uses request->masterXaction to detect IP addresses as per equivalents mentioned by Amos in link above.

And overall logic remains same.

## Why not use AccessLogEntryPointer as parameter?
I think having:
```const Ip::Address *FindListeningPortAddress(const HttpRequest *request, const Ip::Address *cacheip, const Ip::Address *localip);```

is better than having:

```const Ip::Address *FindListeningPortAddress(const HttpRequest *request, const AccessLogEntry *al);```

Because if tomorrow - if we need to use FindListeningPortAddress() somewhere else (where different class is available) - we can not keep adding that class.

For eg:
```const Ip::Address *FindListeningPortAddress(const HttpRequest *request, const AccessLogEntryPointer &ale, const FooClass *bar);```

And then we will have to add code that will access the bar->tcpClient or bar->cache.port too.

So best is that calling function provides cacheip and localip as parameters which is all we need. And if calling function does not know cacheip or localip - then code will use request->masterXaction to detect the same.

OR we can add one more parameter dont_use_masterXaction in which case missing IPs will not be derived from request->masterXaction.

Reply:
First of all, thank you for explaining your rationale! Such explanations are rare (unfortunately) and make the discussion much more productive. 

> In my opinion my new code FindListeningPortAddress() is ...

Please note that I was not attacking your implementation of that function. I was only attacking its interface. I will re-review its implementation once the interface is settled again.

> its impossible for me to go through whole code and prove that masterXaction->squidPort is equivalent to al->cache.port

If Amos thinks they are equivalent (in this PR context), and you see no reasons to think otherwise, then that is good enough for me! They _should be_ equivalent, of course.

We should  document the assumptions in FindListeningPortAddress() implementation. Your assumptions might be different, but I hope they are close to this example (which is based on how the code should work):

```
// Assume al->cache.port and masterXaction->squidPort are equivalent (but only
// one may be available). Same for al->tcpClient and masterXaction->tcpClient.
```

> tomorrow - if we need to use FindListeningPortAddress() somewhere else (where different class is available) - we can not keep adding that class.

I see several problems in the above logic:

1. The biggest problem with adding parameters that are likely to require `foo->bar->boo->baz` calculations in the caller is that developers sooner or later will crash Squid by forgetting to check that foo, baz, and boo are not nil. That has happened many times! The `baz` parameter itself is not the biggest problem. And when all those leaf parameters are rooted in an ALE class designed for sharing transaction information, we should be passing an ALE pointer if at all possible instead of dereferencing it five times in the caller to compute the leaf nodes.
1. The core information needed by FindListeningPortAddress() should be in the master transaction (and nowhere else). Currently, Squid suffers from master transaction being essentially represented by 2.5 classes (HttpRequest/MasterXaction and AccessLogEntry/HttpRequest). We should fix that, of course, but that duality explains why it is natural to use two parameters (for now) and also explains why it is somewhat unlikely that somebody will need to add a third parameter representing the same core master transaction info.
1. The need for a third parameter is a speculation, but arrival of ALE in error pages is actually [anticipated](https://github.com/measurement-factory/squid/pull/4). Supporting an ALE parameter now will avoid code duplication or rewrites when ALE is added to error pages.
1. If adding a third parameter representing core master transaction info becomes necessary, we can split FindListeningPortAddress() into public functions that have easy/safe-to-compute parameters and a low-level function that does the actual work, hiding the dangers from the callers and still requiring them to supply the right parameters.
1. Default parameters are the wrong solution for passing information that is required when it is available. Sooner or later, a developer will copy-paste the call from the wrong context (where the optional parameters were justifiably omitted) and the reviewer will not notice that the (invisible) parameters are required in the new context. This has happened many times!
1. The two-parameter solution is much simpler and does not require the caller to understand whether "IPs must be derived". Frankly, I am not even sure why the caller would _not_ want those IPs to be "derived" if derivation is needed and possible (which the function itself can decide without caller help).

Reply:
In summary, given your explanations, I recommend the second interface option (with two parameters) and with whatever internal FindListeningPortAddress() implementation you think is best.